designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -72,77 +72,134 @@ class DeploymentServiceImpl(
+              .deploymentManagerUnsafe(ctx.processDetails.processingType)
+              .processCommand(command)
+          }
+          Future.successful(deploymentFuture)
       }
-    }
+    } yield result
+  }
+
+  /**
+    * Common validations and operations for a command execution.
+    * @return gathered data for further command execution
+    */
+  private def prepareCommandContextWithAction[PS: ScenarioShapeFetchStrategy](
+      processId: ProcessIdWithName,
+      actionType: ProcessActionType,
+      getVersionOnWhichActionIsDone: ScenarioWithDetailsEntity[PS] => Option[VersionId],
+      getBuildInfoProcessingType: ScenarioWithDetailsEntity[PS] => Option[ProcessingType]
+  )(implicit user: LoggedUser, ec: ExecutionContext): Future[CommandContext[PS]] = {
+    implicit val freshnessPolicy: DataFreshnessPolicy = DataFreshnessPolicy.Fresh
+    val actionName                                    = ScenarioActionName(actionType)
+    dbioRunner.runInTransaction(
+      for {
+        // 1.1 TODO common parameter validations
+        // 1.2 lock for critical section
+        _ <- actionRepository.lockActionsTable
+        // 1.3. fetch scenario data
+        processDetailsOpt <- processRepository.fetchLatestProcessDetailsForProcessId[PS](processId.id)
+        processDetails    <- existsOrFail(processDetailsOpt, ProcessNotFoundError(processId.name))
+        // 1.4. calculate which scenario version is affected by the action: latest for deploy, deployed for cancel
+        versionOnWhichActionIsDone = getVersionOnWhichActionIsDone(processDetails)
+        buildInfoProcessingType    = getBuildInfoProcessingType(processDetails)
+        // 1.5. check if action is performed on proper scenario (not fragment, not archived)
+        _ = checkIfCanPerformActionOnScenario(actionName, processDetails)
+        // 1.6. check if action is allowed for current state
+        inProgressActionTypes <- actionRepository.getInProgressActionTypes(processDetails.processId)
+        processState          <- getProcessState(processDetails, inProgressActionTypes)
+        _ = checkIfCanPerformActionInState(actionName, processDetails, processState)
+        // 1.7. create new action, action is started with "in progress" state, the whole command execution can take some time
+        actionId <- actionRepository.addInProgressAction(
+          processDetails.processId,
+          actionType,
+          versionOnWhichActionIsDone,
+          buildInfoProcessingType
+        )
+      } yield CommandContext(processDetails, actionId, versionOnWhichActionIsDone, buildInfoProcessingType)
+    )
   }
 
+  // TODO: Use buildInfo explicitly instead of ProcessingType-that-is-used-to-calculate-buildInfo
+  private case class CommandContext[PS: ScenarioShapeFetchStrategy](
+      processDetails: ScenarioWithDetailsEntity[PS],
+      actionId: ProcessActionId,
+      versionOnWhichActionIsDone: Option[VersionId],
+      buildInfoProcessingType: Option[ProcessingType]
```

Comment:
Why not just use `ScenarioWithDetailsEntity.processingType`?

Reply:
So far for `deploy` we have `ScenarioWithDetailsEntity.processingType`, for `cancel` we have `None`. I guess this is similar to the "VersionOnWhichActionIsDone", and that we want to keep track of buildInfo of deployed scenario. But... processingType does not change and we want to keep buildInfo for the point in time when action is invoked.

Reply:
I think that we can try to break compatibility and store buildInfo / version for every action but we can do this later